Filtering of old ext data imports + major import estimates code cleanup

poltak opened the pull request:
Worked on this for the last few days originally to get the same filtering (blacklisting, loggability, URL normalisation) on old ext data, but ended up cleaning up and restructuring the way imports module constructs its TODO list. 

Managed to reduce the calculation time _on my data_ to ~3s, down from 6s by changing the order of some filters and ensuring certain expensive things only happen as much as they need to. Also a bunch of logical bugs fixed that were affecting some of the counts under different circumstances.

Still could be some weird cases, so play around and report if you see them. Bookmarks and history now are completely separate to old ext items, as it was causing so much issues previously. Old ext imports should no longer affect remaining counts of bookmarks/history.

A further idea:
We probably don't need to calculate this stuff every single time the user visits imports. Maybe we do it the first time, then store that value for reuse until a certain timeout (every day?). It would improve the UX of the UI a lot (as it's just plain annoying now) for the sake of a bit of accuracy. When they press "start import" we would recalculate.

oliversauter commented:
> We probably don't need to calculate this stuff every single time the user visits imports. Maybe we do it the first time, then store that value for reuse until a certain timeout (every day?). It would improve the UX of the UI a lot (as it's just plain annoying now) for the sake of a bit of accuracy. When they press "start import" we would recalculate.

I like this idea of recalculating before a user starts. This makes at least sure we won't forget any items.

Another idea: 
- We could recalculate if a person changed the blacklist filter since the last estimate, which always has a major effect. (and it actually feels nice to see the items go down) :D

oliversauter commented:
> Bookmarks and history now are completely separate to old ext items

What implications does that have in terms of duplicates?

poltak commented:
> What implications does that have in terms of duplicates?

If a user chooses to do history/bookmarks import first or browses around before ever doing imports, the old ext import counts should not change. Later when they run old ext import, if the page is already indexed, it will reindex it.

How about if we do a check during the import process (not the estimates/import-items creation) to ignore the page part if page exists and is recent? In this case, we would instead just convert the visit + bookmark, skipping the main time-consuming XHR. Should be able to speed up old ext imports a bit in cases where the user browses around a bit or plays with other imports first.

obsidianart commented on src/imports/background/import-item-creation.js:
```diff
@@ -0,0 +1,195 @@
+async function checkWithExistingPages() {
+    const { rows: existingPageDocs } = await db.allDocs({
+        startkey: pageKeyPrefix,
+        endkey: `${pageKeyPrefix}\uffff`,
```
what is uffff?

poltak replied:
This is the highest 2-byte unicode character ￿. Basically it's saying "all docs with key prefix `page/`"
Used in Pouch range queries like this as Pouch keeps things sorted internally by doc IDs, and this character will have a higher sort order than anything else we store. Maybe some more info in the [Pouch docs here](https://pouchdb.com/api.html#batch_fetch) (they actually use the character `\ufff0` here. Could you think of any difference in behaviour?)

obsidianart replied:
probably not, I was just curious

obsidianart commented on src/imports/background/import-item-creation.js:
```diff
@@ -0,0 +1,195 @@
+                const encodedUrl = encodeUrl(currItem.url, false)
+
+                // Filter out items already existing in DB, if specified
+                if (alreadyExists == null || !alreadyExists(encodedUrl)) {
```
the alreadyExists check can be probably move on line 76, non need to enter the try catch and encode the url for it

poltak replied:
`encodeUrl` should throw a malformed URI error for certain URLs that have placeholders in them. Like `%s` in `http://www.youtube.com/results?search_query=%s` commonly used in keyword search bookmarks), hence the `try...catch` here.

obsidianart replied:
I understand, but you can reduce the try catch scope to encodeUrl only instead of the entire block

obsidianart commented on src/imports/background/import-item-creation.js:
```diff
@@ -0,0 +1,195 @@
+                }
+                return accItems
+            } catch (error) {
+                // URI malformed; don't add to list
```
is this the catch only for encodeUrl or the transform can fail too?

obsidianart replied:
I generally don't care much about sub-optimization but since this is a refactor for speed it's worth trying to reduce the try-catch size (https://github.com/petkaantonov/bluebird/wiki/Optimization-killers)

poltak replied:
This article looks interesting 👍 only the `encodeUrl` can fail. `transform` is just a pure function mapping source data to the imports data shape, or identity function by default.

obsidianart replied:
as above

obsidianart commented on src/imports/background/import-item-creation.js:
```diff
@@ -0,0 +1,195 @@
+        //  (eg.I indexed 400 pages, index contained 43 million)
+        const indexSet = new Set(index.index)
+
+        for (const keyChunk of chunk(200)([...indexSet])) {
```
I have no idea of the performance of chunk and the spread operator (I assume quite low compared to a forloop ending at 200.

obsidianart replied:
Reading the core more I'm not sure I understand the reason for this chunking in the first place, and what ```browser.storage.local.get(keyChunk)``` contains?

poltak replied:
This is the old extension data. Each page/visit was stored as its own key in local storage. There is a special key `index` which is an array of all the other keys, acting as the reverse index. 

So here we're grabbing `index` array, breaking it into chunks, and then accessing that chunk of 200 docs from local storage, transforming it to new structure, before moving onto next chunk. Currently splitting it into chunks with the thoughts (both relative to the # of old ext docs):
 - put a constant cap on space
 - give a `N / #chunks` speedup in time (compared to `N` if we access one-at-a-time)

Could be replicated with old-style for-loop, although that would just be another sub-optimization, right? Might as well give it a try

obsidianart replied:
if this is a one-off from the old index to the new one probably is not worth spending more time that you did

obsidianart commented on src/imports/background/import-item-creation.js:
```diff
@@ -0,0 +1,195 @@
+        const indexSet = new Set(index.index)
+
+        for (const keyChunk of chunk(200)([...indexSet])) {
+            const storageChunk = await browser.storage.local.get(keyChunk)
```
a blocking operation inside a loop seems a bad idea, I wonder if we can change this to parallel (I would need some testing about performances), something like (I'm sorry if the syntax is incorrect, I generally optimize for readability not speed)
```
const chunk200 =chunk(200)([...indexSet]
Promise.all(
	chunk200
    //assuming browser etc is a chainable promise, otherwise the standard new Promise etc
    .map(keyChunk=>Promise.resolve(browser.storage.local.get(keyChunk))) 
)
.then([...storageChunk]=>{
    for (const [index,keyChunk] of chunk200.entries()) {
	    keyChunk.forEach((key, i) => {
	        if (storageChunk[index][key] == null) {
	            return
	        }
	        importItems.push(transform(storageChunk[index][key], i))
	    }) 
    }
}
```
On the other side, I have a suspicion this code did intentionally use await to don't be blocking, is it the case?

poltak replied:
Yes intentionally blocking. Common way to cap the space complexity to a constant level, tradeoff being more time. 

If we use `Promise.all` here, each `browser.storage.local.get` call would be processing in "parallel", but our space usage here would be linear to the old ext index size. We had this like this until last week, then we released and it blew up everyone's browser who had sufficient amount of old ext data.

obsidianart replied:
got it. Probably worth adding a comment in the code to explain why you are doing it like that

obsidianart commented on src/imports/background/import-item-creation.js:
```diff
@@ -0,0 +1,195 @@
+
+    // Get all history from browser, filter on existing DB pages
+    let historyItemsMap = filterItemsByUrl(
+        await getHistoryItems(),
```
if you have 2 await is probably better to use parallel promises (if this function is not called within the loop it's not very relevant)

poltak commented:
Thanks for taking the time to review my code @obsidianart! Really appreciate it 😄

obsidianart commented:
my pleasure, quite an interesting piece to read

poltak commented:
Implemented the caching of estimate calculations. Works by using a past-calculation's results if they are less than a day old. Re-calculation happens automatically on import start and cancellation to ensure an import has the latest data. Will add in @oliversauter's suggestion of forcing recalc on blacklist change in currently ongoing blacklist work (just needs to dirty the stored calc's timestamp when adding new item).

Added in @obsidianart's suggestion of reducing the `try...catch` scope and allowing some Promises to run at the same time. Also played with some micro-optimisation with the loops. Changing the main filtering `Array.prototype.reduce` call to a traditional indexed `for` loop gave ~300ms speedup for my data, which is nice! 

Also reduced the space complexity of calculating browser history items by doing it in batches of 2 weeks (previously was the processing the entire available 3 month period at once).

All up calculating my data has reduced from ~6s to ~2.5s through all the changes in this PR. And fixed a few bugs related to counts.

obsidianart commented:
By reading more try catch should be in a separate function or the entire function can not be optimized (worth trying https://github.com/petkaantonov/bluebird/wiki/Optimization-killers)

poltak commented:
@obsidianart yes I noticed that babel does a similar thing via regenerator when compiling down `try...catch` within `async` contexts. Added this in manually in d742951 and, although I don't notice any difference, going by that article it should be a bit of a boost in older Chromes, especially as it can be run 10-20k times depending on their history. A bit of googling tells me SpiderMonkey seems to handle it fine.

oliversauter commented:
Tested the workflows today with my 12k documents. 
I realized I never tested this properly before it was merged. My bad. 

The workflow is still a bit rocky and the numbers are somehow off.
Also the filtering of old extension data also does not work. 

In fact, changing the blacklist does not affect any of the numbers, when i go to the import page right after, so I assume no recalc happens when the import page is opened again. 
Also when I come back from just importing a few documents, the change does not reflect the number of documents I just processed. When I have processed 2, it sometimes changes the numbers to 4 up/down respectively. It always has 2 items too much (so download 5, change by 7)

Also in this workflow, we have actually not improved the workflow of recalculating. 
The problem before was that we needed to calculate when we opened the import page, then when we finished it up, and then everytime we need to open it again. 
Now this recalc happens, whenever we finish a download and come back to the main view, which is basically the same. 

Maybe it is better to change the estimate count with each processed document, so we dont have to recalc in this work-flow so often.  
So basically with a successful entry of a page, the number of to-be-imported documents should go down, the number of documents in the db go up, same for bookmarks and the respective failed ones.  
Forcing to recalc we only need to do on blacklist change or if a person activated the pause button at some point.

poltak commented:
> changing the blacklist does not affect any of the numbers, when i go to the import page right after

This will be coming in #194. On this branch it should still do it when it recalcs on import start, just the estimates view won't update yet unless your stored calculations expires.

> Also the filtering of old extension data also does not work.

Filtering using blacklist, right? I presume this is as a result of the same difference in expected blacklist behaviour, as this one is working great for me. 

> Also when I come back from just importing a few documents, the change does not reflect the number of documents I just processed. When I have processed 2, it sometimes changes the numbers to 4 up/down respectively. It always has 2 items too much (so download 5, change by 7)

This one is a known bug since we first implemented imports, unrelated to this work. Won't be able to properly fix until we implement cancellable Promise support in the imports ([talked about here](https://github.com/WorldBrain/Memex/issues/143#issuecomment-343710804)).

> The problem before was that we needed to calculate when we opened the import page, then when we finished it up, and then every time we need to open it again.
Now this recalc happens, whenever we finish a download and come back to the main view, which is basically the same.

This solved the "when we opened import page" one, which is a huge improvement as it effects users not even running an import.
The calc that happens on start press should generally always be run to get the most up-to-date data, but I think we can skip it if we determine the time from the last calc is within some time (say 10 mins?) - useful if they come to the page, calc runs, then they start import.
The suggestion with altering the pre-calc'd estimate counts as imports is running seems like a good one. It will be more complicated than +/-1 each doc however as there are inter-dependencies between the numbers.
